### PR TITLE
Add initial support for GPTQ

### PR DESCRIPTION
**Pull request by WoosukKwon**
This PR is a simplified version of the great PR #916.
The main difference is that, this PR does not use the exllama kernels while #916 does.
The purpose of this PR is to minimize the code change in a PR, and avoid possible conflicts from @zhuohan123's ongoing refactoring effort.

### CODE REVIEW

**Comment by LimpidEarth** [comment]
Excited to see this PR. According to "Files Changed", it seems that this PR only support 4 bit quantization. How about 8 bit quantization? Is there a specific support plan? Thanks

**Comment by zhaoyang-star** [comment]
Great work! Could you please share the performance gained compared the float16?

**Comment by WoosukKwon** [comment]
@zhuohan123 The PR is ready for review. Please take a look!

**Comment by WoosukKwon** [comment]
@zhaoyang-star The kernels used in this PR is not optimized, so actually you cannot get any speedup for now. We will optimize the quantized GEMM kernels in the next PR.

@LimpidEarth The kernel we are using now only supports 4 bit quantization, but we can extend it in the next PR. BTW, it seems most of the GPTQ models found in HF model hub are using 4 bit quantization. Could you provide us with any pointer to 8-bit GPTQ model?

**Comment by bash99** [comment]
> @zhaoyang-star The kernels used in this PR is not optimized, so actually you cannot get any speedup for now. We will optimize the quantized GEMM kernels in the next PR.
> 
> @LimpidEarth The kernel we are using now only supports 4 bit quantization, but we can extend it in the next PR. BTW, it seems most of the GPTQ models found in HF model hub are using 4 bit quantization. Could you provide us with any pointer to 8-bit GPTQ model?

Most new GPTQ model released by TheBloke now has 8bit version (-1g 128g 32g)，
https://huggingface.co/TheBloke/OpenHermes-2.5-Mistral-7B-GPTQ/tree/gptq-8bit-128g-actorder_True

File: vllm/config.py [hunk 1]
```diff
@@ -103,15 +103,31 @@ def _verify_tokenizer_mode(self) -> None:
...
+                    f"({self.quantization}).")
+
+        if self.quantization is not None:
+            if self.quantization not in supported_quantization:
+                raise ValueError(
+                    f"Unknown quantization method: {self.quantization}. Must "
+                    f"be one of {supported_quantization}.")
+        logger.warning(f"{self.quantization} quantization is not fully "
+                       "optimized yet. The speed can be slower than "
+                       "non-quantized models.")
```
Comment (zhuohan123) [inline vllm/config.py:130]: Does this mean all quantization methods are not optimized yet?

File: vllm/model_executor/layers/quantized_linear/gptq.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+                device="cuda",
+                dtype=torch.int32,
+            ),
+            requires_grad=False,
+        )
+        self.shifter = torch.tensor(
+            [0, 4, 8, 12, 16, 20, 24, 28],
+            device="cuda",
+            dtype=torch.int32,
+        )
```
Comment (zhuohan123) [inline vllm/model_executor/layers/quantized_linear/gptq.py:94]: Is this `shifter` a must-have? The problem here is that we created a "non-parameter" tensor. We will need to modify the weight creation code in #1622 to make this creation work.

File: vllm/model_executor/layers/quantized_linear/gptq.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+    def apply_weights(
+        self,
+        x: torch.Tensor,
+        bias: Optional[torch.Tensor],
+    ) -> torch.Tensor:
+        out_shape = x.shape[:-1] + (self.qweight.shape[-1], )
+        reshaped_x = x.reshape(-1, x.shape[-1])
+        num_tokens = x.shape[:-1].numel()
+        if num_tokens <= 32:
```
Comment (zhuohan123) [inline vllm/model_executor/layers/quantized_linear/gptq.py:104]: Why do we have an `if` here? Is the CUDA kernel slower when `num_tokens > 32` or it's just the CUDA kernel will not work at all?

File: vllm/model_executor/layers/quantized_linear/gptq.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+                self.input_size // group_size,
+                self.output_size,
+                device="cuda",
+                dtype=dtype,
+            ),
+            requires_grad=False,
+        )
+        # Initialize g_idx to be sequential.
+        # This is required because old GPTQ models may not have g_idx.
+        start_idx = self.tp_rank * self.input_size_per_partition
```
Comment (zhuohan123) [inline vllm/model_executor/layers/quantized_linear/gptq.py:157]: Just leave as a note: This line will also require some modification in #1622.

**Review by zhuohan123 — COMMENTED** [review]
Thanks for the work! The changes in general LGTM. However, there are some places that will conflict with #1622. Mainly because of the auxiliary variables like `g_idx` and `shifter`. What do you think about the merging plan? Should we merge this first or #1622 first? No matter which plan we go, I can help on the merge.

File: vllm/model_executor/layers/quantized_linear/gptq.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/model_executor/layers/quantized_linear/gptq.py:94]: How difficult will it be to add support for such tensors? While this tensor is only used for the pytorch-based GPTQ matmul implementation and thus will be eventually unused once we develop a more optimized kernel, such non-parameter buffers can be used for other quantization methods. I believe we should take this into account in the new design.

File: vllm/model_executor/layers/quantized_linear/gptq.py (hunk 3 above)
Comment (WoosukKwon) [reply vllm/model_executor/layers/quantized_linear/gptq.py:104]: Good point. Actually, the current GPTQ kernel is designed for batch size 1 and performs extremely poorly when the batch size is large, often taking 10+ minutes for the initial memory profiling. As a workaround, I implemented a simple PyTorch-based GPTQ matmul faster than the original kernel for large batch size. Still, the two implementations are very bad and probably much slower than the optimized implementations like exllama.

File: vllm/config.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/config.py:130]: Unfortunately, yes. For the three quantization methods we support, we are using the original authors' kernels, which can be further optimized. Particularly, the squeezellm and GPTQ kernels are slow if batch size > 1. As for AWQ, I think its kernel is much better than the two, but is still slow for large batch size and does not support bfloat16.

File: vllm/model_executor/layers/quantized_linear/gptq.py (hunk 3 above)
Comment (WoosukKwon) [reply vllm/model_executor/layers/quantized_linear/gptq.py:108]: Added a comment on this.

**Comment by LimpidEarth** [comment]
> @LimpidEarth The kernel we are using now only supports 4 bit quantization, but we can extend it in the next PR. BTW, it seems most of the GPTQ models found in HF model hub are using 4 bit quantization. Could you provide us with any pointer to 8-bit GPTQ model?

Got it and looking forward to the PR of 8 bit supporting! The main reason for 8bit GPTQ model is that we found the evaluation results are better than 4 bit model regarding our domain tasks.

**Comment by WoosukKwon** [comment]
Closed as some of the changes are already merged and #916 will be merged instead.

**Pull request closed without merging**